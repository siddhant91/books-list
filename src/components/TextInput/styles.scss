@import '../../styles/main.scss';
.bokl-text-input {
	position: relative;
	display: block;
	max-width: 345px;
	margin-bottom: $bokl-spacing-5;
	&--label {
		border: none;
		border-bottom: 1px solid rgba($bokl-color-black, 0.32);
		padding: $bokl-spacing-2 0;
	}
	input {
		appearance: none;
		font-size: $bokl-body-font-size;
		width: 100%;
		box-sizing: border-box;
		background: $bokl-color-white-1;
		border: none;
		outline: none;
		font-weight: $bokl-font-weight-normal;
		margin: 0 $bokl-spacing-1;

		/* clears the 'X' from Internet Explorer */
		&::-ms-clear,
		&::-ms-reveal {
			display: none;
			width: 0;
			height: 0;
		}

		/* clears the 'X' from Chrome */
		&::-webkit-search-decoration,
		&::-webkit-search-cancel-button,
		&::-webkit-search-results-button,
		&::-webkit-search-results-decoration {
			display: none;
		}
	}

	&--clear-icon {
		position: absolute;
		top: 0;
		bottom: 0;
		right: 0;
		color: $bokl-color-graphite;
		margin-right: 5px;
		cursor: pointer;
		display: flex;
		align-items: center;
	}
	&--error-text {
		color: $bokl-color-red;
	}
}

// .text-input-component .clear > .clear__icon {
// 	display: flex;
// 	justify-content: center;
// 	align-items: center;
// 	border: solid 1px;
// 	border-radius: 50%;
// 	height: 14px;
// 	width: 14px;
// 	padding: 0 0 2px 1px;
// }
