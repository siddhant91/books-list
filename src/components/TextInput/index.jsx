import clsx from 'clsx';
import { bool, node, oneOf, string } from 'prop-types';
import { useRef } from 'react';
import { useField, useFormikContext } from 'formik';

// Styles
import './styles.scss';

function TextInput({
	additionalClasses,
	children,
	readOnly,
	type,
	ariaLabel,
	canClear,
	placeHolder,
	labelKey,
	...props
}) {
	const textInput = useRef(null); //  reference to the input field
	const [field, meta] = useField(props);
	const { value = '' } = field;
	const { touched, error } = meta;
	const isInValid = touched && error;
	const { setFieldValue } = useFormikContext();

	/**
	 * Triggered on click of CLEAR icon.
	 * Clears the input field and sets the focus on the input field.
	 */
	const clearField = () => {
		setFieldValue(field.name, '');
		textInput.current.focus();
	};

	return (
		<div
			className={clsx('bokl-text-input', additionalClasses, {
				disabled: readOnly,
			})}
		>
			<label htmlFor={labelKey} className="d-flex bokl-text-input--label">
				<span>
					<i className="fa fa-search" aria-hidden="true" />
				</span>
				<input
					id={labelKey}
					type={type}
					value={value}
					aria-label={ariaLabel}
					data-testid="TextInput"
					ref={textInput}
					className={clsx({
						disabled: readOnly,
					})}
					{...field}
					{...props}
				/>
			</label>
			{
				/**
				 * Show CLEAR (X) icon only when when
				 * `canClear` is true (default)
				 * AND the input is not blank
				 * AND the field is NOT read-only/disabled.
				 */
				canClear && value.length > 0 && !readOnly && (
					<div
						className="bokl-text-input--clear-icon"
						role="button"
						tabIndex="0"
						aria-pressed="false"
						aria-label="clear-TextInput"
						onClick={clearField}
						onKeyPress={(e) => {
							if (e.key === 'Enter') {
								clearField();
							}
						}}
						data-testid="clear-TextInput"
					>
						<i className="fa fa-times" aria-hidden="true" />
					</div>
				)
			}
			{isInValid && <p className="bokl-text-input--error-text">{error}</p>}
			{children}
		</div>
	);
}

TextInput.propTypes = {
	children: node,
	readOnly: bool,
	additionalClasses: string,
	type: oneOf(['text', 'email', 'password', 'number', 'tel', 'search']),
	ariaLabel: string,
	canClear: bool,
	placeHolder: string,
	labelKey: string.isRequired,
};

TextInput.defaultProps = {
	children: null,
	readOnly: false,
	additionalClasses: '',
	type: 'search',
	ariaLabel: 'TextInput',
	canClear: true,
	placeHolder: '',
};

export default TextInput;
