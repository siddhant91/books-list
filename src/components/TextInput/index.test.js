import { render, fireEvent, cleanup } from '@testing-library/react';

import { Formik, Form } from 'formik';

import TextInput from './index';

afterEach(cleanup);

const renderComponent = ({ initialSearchText = '', readOnly = false } = {}) => {
	return (
		<Formik
			initialValues={{
				searchText: initialSearchText,
			}}
			onSubmit={(values) => {
				console.log(values.searchText);
			}}
		>
			{() => (
				<Form autoComplete="off">
					<TextInput
						labelKey="Search For Books"
						name="searchText"
						placeholder="Search"
						readOnly={readOnly}
					>
						<p>Text</p>
					</TextInput>
				</Form>
			)}
		</Formik>
	);
};
it('Matches snapshot', () => {
	const { asFragment } = render(renderComponent());
	expect(asFragment()).toMatchSnapshot();
});

it('Renders children props', () => {
	const { getByText } = render(renderComponent());
	expect(getByText('Text')).toBeInTheDocument();
});

it('Renders value props correctly', () => {
	const { getByTestId } = render(renderComponent({ initialSearchText: 'testSearch' }));
	expect(getByTestId('TextInput').value).toBe('testSearch');
});

it('should disable the TextInput', () => {
	const { getByTestId } = render(renderComponent({ readOnly: true }));
	expect(getByTestId('TextInput').readOnly).toBeTruthy();
});

it('Handles change in TextInput input element', async () => {
	const { getByTestId, findByTestId } = render(renderComponent({}));
	const element = getByTestId('TextInput');
	fireEvent.change(element, { target: { value: 'Value' } });
	const newInput = await findByTestId('TextInput');
	expect(newInput.value).toBe('Value');
});

it('Handles canClear button interactions', async () => {
	const { getByTestId, findByTestId } = render(
		renderComponent({ initialSearchText: 'testSearch' }),
	);
	const element = getByTestId('clear-TextInput');
	fireEvent.click(element);
	const newInput = await findByTestId('TextInput');
	expect(newInput.value).toBe('');
});

it('Handles canClear button interactions with Enter Key', async () => {
	const { getByTestId, findByTestId } = render(
		renderComponent({ initialSearchText: 'testSearch' }),
	);
	const element = getByTestId('clear-TextInput');
	fireEvent.keyPress(element, { key: 'Enter', keyCode: 13 });
	const newInput = await findByTestId('TextInput');
	expect(newInput.value).toBe('');
});
