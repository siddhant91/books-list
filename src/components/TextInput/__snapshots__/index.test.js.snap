// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    autocomplete="off"
  >
    <div
      class="bokl-text-input"
    >
      <label
        class="d-flex bokl-text-input--label"
        for="Search For Books"
      >
        <span>
          <i
            aria-hidden="true"
            class="fa fa-search"
          />
        </span>
        <input
          aria-label="TextInput"
          class=""
          data-testid="TextInput"
          id="Search For Books"
          name="searchText"
          placeholder="Search"
          type="search"
          value=""
        />
      </label>
      <p>
        Text
      </p>
    </div>
  </form>
</DocumentFragment>
`;
